src/DDTrace/Integrations/SymfonyMessenger/SymfonyMessengerIntegration.php

```diff
@@ -0,0 +1,348 @@
+<?php
+
+namespace DDTrace\Integrations\SymfonyMessenger;
+
+use DDTrace\HookData;
+use DDTrace\Integrations\Integration;
+use DDTrace\SpanData;
+use DDTrace\Tag;
+use DDTrace\Util\ObjectKVStore;
+use Symfony\Component\Messenger\Bridge\AmazonSqs\Transport\AmazonSqsReceivedStamp;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Handler\HandlerDescriptor;
+use Symfony\Component\Messenger\Handler\HandlersLocatorInterface;
+use Symfony\Component\Messenger\Stamp\BusNameStamp;
+use Symfony\Component\Messenger\Stamp\ConsumedByWorkerStamp;
+use Symfony\Component\Messenger\Stamp\DelayStamp;
+use Symfony\Component\Messenger\Stamp\HandledStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+use Symfony\Component\Messenger\Stamp\RedeliveryStamp;
+use Symfony\Component\Messenger\Stamp\SentStamp;
+use Symfony\Component\Messenger\Stamp\TransportMessageIdStamp;
+use Symfony\Component\Messenger\Transport\Receiver\ReceiverInterface;
+use function DDTrace\hook_method;
+use function DDTrace\install_hook;
+use function DDTrace\remove_hook;
+use function DDTrace\trace_method;
+
+class SymfonyMessengerIntegration extends Integration
+{
+    const NAME = 'symfonymessenger';
+
+    public function init(): int
+    {
+        $integration = $this;
+
+        if (!\class_exists(\Symfony\Component\Messenger\Event\WorkerStartedEvent::class)) {
+            // Only exists in Symfony Messenger 4.4+
+            return Integration::NOT_LOADED;
+        }
+
+        trace_method(
+            'Symfony\Component\Messenger\MessageBusInterface',
+            'dispatch',
+            function (SpanData $span, array $args, $retval, $exception) use ($integration) {
+                $integration->setSpanAttributes($span, 'symfony.messenger.dispatch', null, $retval ?? $args[0], null, null, true);
+                if ($exception) {
+                    // Worker::handleMessage() will catch the exception. We need to manually attach it to the root span.
+                    \DDTrace\root_span()->exception = $exception;
+                }
+            }
+        );
+
+        // Attach current context to Envelope before sender sends it to remote queue
+        install_hook(
+            'Symfony\Component\Messenger\Transport\Sender\SenderInterface::send',
+            function (HookData $hook) {
+                /** @var \Symfony\Component\Messenger\Envelope $envelope */
+                $envelope = $hook->args[0];
+
+                if (\ddtrace_config_distributed_tracing_enabled()) {
```

Comment:
Should this be governed by `DD_TRACE_SYMFONY_MESSENGER_DISTRIBUTED_TRACING`? Which triggers another question: Why add a new setting for this instead of using the global distributed tracing setting? Do we expect customers to have it enabled for one and not the other? Is it just for better customization?

Reply:
`DD_TRACE_SYMFONY_MESSENGER_DISTRIBUTED_TRACING` will state whether the produce <-> consume relationship happens in the traditional distributed tracing manner - i.e., within the same trace - or if it should leverage span links. This decision is made on the consuming end of the operation.

In retrospect, `ddtrace_config_distributed_tracing_enabled` will state whether the produce <-> consume relationship will exist or not. This decision is made on the producing end of the operation.

We expect customers to have the global distributed tracing setting enabled. Regarding `DD_TRACE_SYMFONY_MESSENGER_DISTRIBUTED_TRACING`, we also expect it to be enabled by default - i.e., by default, we don't change the UX customers are used to. However, we offer the possibility to use span links instead, which has proved to be particularly effective to un-bloat flamegraphs. In the future, this could even be the default behavior... but that's TBD.

This comment just reminded me that I've to update the public doc though.